### PR TITLE
Private Datasets: Update use case: Pushing a dataset

### PR DESCRIPTION
**Pull request by s373r**
## Description

Closes: https://github.com/kamu-data/kamu-cli/issues/728

### Added
- Private Datasets, changes related to Smart Transfer Protocol:
  - `kamu push`: added `--visibility private|public` argument to specify the created dataset visibility
  - Send the visibility attribute in the initial request of the push flow
### Changed
- Smart Transfer Protocol: use `CreateDatasetUseCase` in case of creation at the time of the dataset pulling

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: :warning: Present change SmTP, but it does not affect the current work
    
  - [x] Workspace layout and metadata: ✅ 
    
  - [x] Configuration: ✅
    
  - [x] Container images: ✅
- [x] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [x] [Public documentation](https://github.com/kamu-data/kamu-docs/): ❌ not needed
  
- [x] Downstream effects:
    
  - [x] [kamu-node](https://github.com/kamu-data/kamu-node): ❌ not needed

### CODE REVIEW

File: src/app/cli/src/commands/push_command.rs [hunk 1]
```diff
@@ -167,6 +172,13 @@ impl Command for PushCommand {
             ));
         }
 
+        // todo как будет интернет проверить по рфц, делаем ли мы поддержку для all
+        // и/или рекурсив
```
Comment (zaychenko-sergei) [inline src/app/cli/src/commands/push_command.rs:176]: Упс

File: src/infra/core/src/sync_service_impl.rs [hunk 2]
```diff
@@ -145,12 +144,21 @@ impl SyncServiceImpl {
                     }
                     Err(GetDatasetError::NotFound(_)) if create_if_not_exists => {
                         let alias = local_ref.alias().unwrap().clone();
-                        let repo_writer = self.dataset_repo_writer.clone();
+                        let create_dataset_use_case = self.create_dataset_use_case.clone();
```
Comment (zaychenko-sergei) [inline src/infra/core/src/sync_service_impl.rs:147]: Normally service layer wouldn't invoke a higher level component like use case.
But here, we are rather missing use case layer for push/pull operations.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
Carefully merge with Roman's change in smart protocol, and make sure this breaking change is release with his breaking change in 1 run

File: src/app/cli/src/commands/push_command.rs (hunk 1 above)
Comment (s373r) [reply src/app/cli/src/commands/push_command.rs:176]: shit :smiling_face_with_tear:

Comment (s373r) [reply src/app/cli/src/commands/push_command.rs]: Removed

File: src/infra/core/src/sync_service_impl.rs (hunk 2 above)
Comment (s373r) [reply src/infra/core/src/sync_service_impl.rs:147]: Yes, I had my doubts too, but came to the conclusion that it's acceptable here (in this case)

**Review by zaychenko-sergei — APPROVED** [review]

**Pull request merged**